libraries/chain/backing_store/db_context_rocksdb.cpp

```diff
@@ -0,0 +1,749 @@
+      return prim_key_iter_type { .table_ei = table_ei, .secondary = noop_secondary, .primary = key, .payer = payer};
+   }
+
+   void db_context_rocksdb::swap(int iterator, account_name payer) {
+      primary_iter_store.swap(iterator, noop_secondary, payer);
+   }
+
+   // gets a prefix that allows for only primary key iterators
+   prefix_bundle db_context_rocksdb::get_primary_slice_in_primaries(name code, name scope, name table, uint64_t id) {
+      bytes primary_key = db_key_value_format::create_primary_key(scope, table, id);
+      return { primary_key, end_of_prefix::at_type, code };
+   }
+
+   pv_bundle db_context_rocksdb::get_primary_key_value(name code, name scope, name table, uint64_t id) {
+      prefix_bundle primary_key = get_primary_slice_in_primaries(code, scope, table, id);
+      return { primary_key, current_session.read(primary_key.full_key) };
+   }
+
+   void db_context_rocksdb::set_value(const shared_bytes& primary_key, const payer_payload& pp) {
+      current_session.write(primary_key, pp.as_payload());
+   }
+
+   void db_context_rocksdb::update_db_usage( const account_name& payer, int64_t delta, const storage_usage_trace& trace ) {
+      context.update_db_usage(payer, delta, trace);
+   }
+
+   int32_t db_context_rocksdb::find_and_store_primary_key(const session_type::iterator& session_iter,
+                                                          int32_t table_ei, const shared_bytes& type_prefix,
+                                                          int32_t not_found_return, const char* calling_func,
+                                                          uint64_t& found_key) {
+      // if nothing remains in the database, return the passed in value
+      if (session_iter == current_session.end() || !primary_lookup.match_prefix(type_prefix, (*session_iter).first)) {
+         return not_found_return;
+      }
+      EOS_ASSERT( db_key_value_format::get_primary_key((*session_iter).first, type_prefix, found_key), db_rocksdb_invalid_operation_exception,
+                  "invariant failure in find_and_store_primary_key, iter store found to update but no primary keys in database");
+
+      const account_name found_payer = payer_payload(*(*session_iter).second).payer;
+
+      return primary_iter_store.add(primary_key_iter(table_ei, found_key, found_payer));
+   }
+
+   // returns the exact iterator and the bounding key (type)
+   db_context_rocksdb::exact_iterator db_context_rocksdb::get_exact_iterator(
+         name code, name scope, name table, uint64_t primary) {
+      const auto slice_primary_key = get_primary_slice_in_primaries(code, scope, table, primary);
+      auto session_iter = current_session.lower_bound(slice_primary_key.full_key);
+      const bool valid = primary_lookup.match(slice_primary_key.full_key, session_iter);
+      return { .valid = valid, .itr = std::move(session_iter), .type_prefix = std::move(slice_primary_key.prefix_key) };
```

Comment:
`shared_bytes` is not movable

Reply:
Are you saying it can never be moved, or that it cannot be moved because it is const in this case?

Reply:
Because it is const.